# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [7]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [8]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [9]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [10]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [11]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [12]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [13]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [14]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [15]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [16]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [17]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [18]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [19]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [20]:
df["Season"].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [21]:
df["Wteam"].value_counts()

1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: Wteam, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [22]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [23]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [24]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [25]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [26]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [27]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [28]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [29]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [30]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [31]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [32]:
df.iloc[100:201][["Season","Wteam","Wscore"]]

Season  Wteam  Wscore
100    1985   1409      86
101    1985   1444      90
102    1985   1451      89
103    1985   1139      78
104    1985   1154      78
..      ...    ...     ...
196    1985   1203      83
197    1985   1222      73
198    1985   1227      76
199    1985   1228      73
200    1985   1229      54

[101 rows x 3 columns]

In [33]:
df.sort_values('Season')

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
2484      1985      99   1275      73   1132      63    H      0
2485      1985      99   1306      62   1171      59    H      0
2486      1985      99   1318      99   1145      91    H      0
2487      1985      99   1337      75   1109      56    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
141706    2016      50   1300      57   1183      52    N      0
141705    2016      50   1298      62   1202      49    H      0
141704    2016      50   1295      68   1410      50    N      0
141702    2016      50   1277      99   1324      93    N      1
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

In [34]:
df.sort_values('Wteam')

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0
136740    2015      62   1101      87   1146      70    H      0
143283    2016      89   1101      80   1249      71    H      0
143476    2016      92   1101      79   1223      72    H      0
141972    2016      61   1101      62   1146      61    A      0
...        ...     ...    ...     ...    ...     ...  ...    ...
56143     1998     125   1464      75   1282      59    N      0
94008     2007      41   1464      68   1454      65    A      0
41936     1995      84   1464      67   1282      52    H      0
110589    2010      68   1464      71   1260      67    H      0
46638     1996      98   1464      73   1442      48    H      0

[145289 rows x 8 columns]

In [35]:
df.sort_values('Wscore')

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
128582    2013     116   1264      34   1193      31    H      0
69336     2001     126   1206      35   1423      33    N      0
5364      1986      77   1229      35   1166      34    A      0
20246     1990      47   1374      36   1426      31    H      0
118590    2011     130   1336      36   1458      33    N      0
...        ...     ...    ...     ...    ...     ...  ...    ...
20022     1990      40   1116     166   1109     101    H      0
24341     1991      47   1328     172   1258     112    H      0
19653     1990      30   1328     173   1109     101    H      0
17867     1989      92   1258     181   1109     150    H      0
24970     1991      68   1258     186   1109     140    H      0

[145289 rows x 8 columns]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [38]:
max_score_index = df['Wscore'].idxmax()
max_score_season = df.loc[max_score_index]['Season']
max_score_winning_team = df.loc[max_score_index]['Wteam']

print("Season:", max_score_season)
print("Winning Team:", max_score_winning_team)

Season: 1991
Winning Team: 1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [34]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [39]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [40]:
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

average_difference = df['ScoreDifference'].mean()

filtered_df = df[df['ScoreDifference'] > average_difference]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [41]:
max_difference_index = filtered_df['ScoreDifference'].idxmax()

season = filtered_df.loc[max_difference_index, 'Season']
winning_team = filtered_df.loc[max_difference_index, 'Wteam']
losing_team = filtered_df.loc[max_difference_index, 'Lteam']

print("Season:", season)
print("Winning Team:", winning_team)
print("Losing Team:", losing_team)

Season: 1996
Winning Team: 1409
Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [42]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [43]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [44]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [45]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [48]:
df.groupby('Season')['Wscore'].mean()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [51]:
df.groupby('Wteam')['Wscore'].max()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [55]:
grouped_df = df.groupby(['Season', 'Wteam']).agg({'Wscore': 'mean'})

grouped_df.groupby('Season').apply(lambda x: x[x['Wscore'] == x['Wscore'].max()])

Wscore
Season Season Wteam            
1985   1985   1328    92.800000
1986   1986   1109    91.200000
1987   1987   1380    95.875000
1988   1988   1258   111.750000
1989   1989   1258   117.315789
1990   1990   1258   126.347826
1991   1991   1380   112.312500
1992   1992   1380    99.642857
1993   1993   1380   101.875000
1994   1994   1380   106.583333
1995   1995   1206   102.833333
1996   1996   1206   103.600000
1997   1997   1254    94.142857
1998   1998   1395   101.153846
1999   1999   1317    95.750000
2000   2000   1395    95.647059
2001   2001   1395    99.315789
2002   2002   1242    92.888889
2003   2003   1395    88.222222
2004   2004   1194    92.142857
2005   2005   1311    91.333333
2006   2006   1190    89.533333
2007   2007   1440    99.400000
2008   2008   1440    94.500000
2009   2009   1377    96.200000
2010   2010   1440    95.000000
2011   2011   1449    90.130435
2012   2012   1370    87.400000
2013   2013   1377    86.500000
2014   2014   1322    92.600000
2015   2015   1322    90.733333
2016   2016   1146    97.142857

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [57]:
df.groupby(["Season", "Wteam"]).count()

Daynum  Wscore  Lteam  Lscore  Wloc  Numot  ScoreDifference
Season Wteam                                                             
1985   1102        5       5      5       5     5      5                5
       1103        9       9      9       9     9      9                9
       1104       21      21     21      21    21     21               21
       1106       10      10     10      10    10     10               10
       1108       19      19     19      19    19     19               19
...              ...     ...    ...     ...   ...    ...              ...
2016   1460       20      20     20      20    20     20               20
       1461       12      12     12      12    12     12               12
       1462       27      27     27      27    27     27               27
       1463       21      21     21      21    21     21               21
       1464        9       9      9       9     9      9                9

[10172 rows x 7 columns]

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [58]:
win_count_df = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

win_count_df.loc[win_count_df.groupby('Season')['Wins'].idxmax()]

Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [67]:
ALS_by_season_team = df.groupby(['Season', 'Lteam'])['Lscore'].mean().reset_index()
ALS_by_season_team= ALS_by_season_team.rename(columns={'Lteam': 'Losing Team', 'Lscore': 'Average Losing Score'})

AWS_by_season_team = df.groupby(['Season', 'Wteam'])['Wscore'].mean().reset_index()
AWS_by_season_team =AWS_by_season_team.rename(columns={'Wteam': 'Winning Team', 'Wscore': 'Average Winning Score'})

ALS_by_season_team.merge(AWS_by_season_team, on=['Season'])



Season  Losing Team  Average Losing Score  Winning Team  \
0          1985         1102             61.000000          1102   
1          1985         1102             61.000000          1103   
2          1985         1102             61.000000          1104   
3          1985         1102             61.000000          1106   
4          1985         1102             61.000000          1108   
...         ...          ...                   ...           ...   
3253397    2016         1464             71.238095          1460   
3253398    2016         1464             71.238095          1461   
3253399    2016         1464             71.238095          1462   
3253400    2016         1464             71.238095          1463   
3253401    2016         1464             71.238095          1464   

         Average Winning Score  
0                    71.000000  
1                    70.222222  
2                    72.095238  
3                    75.100000  
4                    85.842105  
...                        ...  
3253397              73.450000  
3253398              73.416667  
3253399              83.222222  
3253400              78.238095  
3253401              85.888889  

[3253402 rows x 5 columns]

In [68]:
merged_df[merged_df['Average Losing Score'] > merged_df['Average Winning Score']]

Season  Losing Team  Average Losing Score  Winning Team  \
5          1985         1102             61.000000          1109   
43         1985         1102             61.000000          1159   
75         1985         1102             61.000000          1201   
157        1985         1102             61.000000          1306   
215        1985         1102             61.000000          1383   
...         ...          ...                   ...           ...   
3253325    2016         1464             71.238095          1385   
3253329    2016         1464             71.238095          1389   
3253342    2016         1464             71.238095          1402   
3253347    2016         1464             71.238095          1407   
3253353    2016         1464             71.238095          1413   

         Average Winning Score  
5                    54.000000  
43                   60.000000  
75                   60.550000  
157                  58.571429  
215                  58.000000  
...                        ...  
3253325              68.142857  
3253329              69.500000  
3253342              70.923077  
3253347              69.857143  
3253353              68.888889  

[274229 rows x 5 columns]

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [69]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [74]:
import numpy as np

df['HighScoringGame'] = 'No'

df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', df['HighScoringGame'])

print(df)



        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3              

Q: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together

In [77]:
team_games = {}

for index, row in df.iterrows():
    winning_team = row['Wteam']
    losing_team = row['Lteam']
    
    if winning_team in team_games:
        team_games[winning_team] += 1
    else:
        team_games[winning_team] = 1

    if losing_team in team_games:
        team_games[losing_team] += 1
    else:
        team_games[losing_team] = 1

for team, num_games in team_games.items():
    print(f"Team {team} played {num_games} games.")



Team 1228 played 992 games.
Team 1328 played 968 games.
Team 1106 played 855 games.
Team 1354 played 906 games.
Team 1112 played 981 games.
Team 1223 played 363 games.
Team 1165 played 833 games.
Team 1432 played 69 games.
Team 1192 played 908 games.
Team 1447 played 903 games.
Team 1218 played 931 games.
Team 1337 played 922 games.
Team 1226 played 847 games.
Team 1242 played 993 games.
Team 1268 played 969 games.
Team 1260 played 914 games.
Team 1133 played 949 games.
Team 1305 played 922 games.
Team 1424 played 974 games.
Team 1307 played 969 games.
Team 1288 played 925 games.
Team 1344 played 951 games.
Team 1438 played 952 games.
Team 1374 played 916 games.
Team 1411 played 903 games.
Team 1412 played 962 games.
Team 1397 played 963 games.
Team 1417 played 966 games.
Team 1225 played 880 games.
Team 1116 played 980 games.
Team 1368 played 808 games.
Team 1120 played 936 games.
Team 1391 played 879 games.
Team 1135 played 847 games.
Team 1306 played 898 games.
Team 1143 played 947 

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [80]:
result_df = pd.DataFrame(columns=['Season', 'Game', 'Score Difference'])

grouped = df.groupby('Season')
for season, group in grouped:
    max_score_diff_index = (group['Wscore'] - group['Lscore']).idxmax()
    max_score_diff = group.loc[max_score_diff_index, 'Wscore'] - group.loc[max_score_diff_index, 'Lscore']
    result_df = result_df.append({'Season': season, 'Game': max_score_diff_index, 'Score Difference': max_score_diff}, ignore_index=True)

for index, row in result_df.iterrows():
    season = row['Season']
    game_index = row['Game']
    score_diff = row['Score Difference']
    print(f"In season {season}, the game with the highest score difference is {game_index} with a score difference of {score_diff}.")


In season 1985, the game with the highest score difference is 236 with a score difference of 60.
In season 1986, the game with the highest score difference is 4731 with a score difference of 84.
In season 1987, the game with the highest score difference is 8240 with a score difference of 73.
In season 1988, the game with the highest score difference is 12046 with a score difference of 68.
In season 1989, the game with the highest score difference is 16677 with a score difference of 70.
In season 1990, the game with the highest score difference is 19502 with a score difference of 76.
In season 1991, the game with the highest score difference is 25161 with a score difference of 68.
In season 1992, the game with the highest score difference is 27997 with a score difference of 82.
In season 1993, the game with the highest score difference is 33858 with a score difference of 81.
In season 1994, the game with the highest score difference is 36404 with a score difference of 69.
In season 1995

C:\Users\Iben\AppData\Local\Temp\ipykernel_14780\3768861180.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  result_df = result_df.append({'Season': season, 'Game': max_score_diff_index, 'Score Difference': max_score_diff}, ignore_index=True)
C:\Users\Iben\AppData\Local\Temp\ipykernel_14780\3768861180.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  result_df = result_df.append({'Season': season, 'Game': max_score_diff_index, 'Score Difference': max_score_diff}, ignore_index=True)
C:\Users\Iben\AppData\Local\Temp\ipykernel_14780\3768861180.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  result_df = result_df.append({'Season': season, 'Game': max_score_diff_index, 'Score Difference': max_score_diff}, ignore_index=T

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [54]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [81]:
games_played = df['Wteam'].value_counts() + df['Lteam'].value_counts()

for team, num_games in games_played.items():
    print(f"Team {team} played {num_games} games.")


Team 1101 played 76 games.
Team 1102 played 840 games.
Team 1103 played 910 games.
Team 1104 played 975 games.
Team 1105 played 447 games.
Team 1106 played 855 games.
Team 1107 played 512 games.
Team 1108 played 866 games.
Team 1109 played 169 games.
Team 1110 played 910 games.
Team 1111 played 876 games.
Team 1112 played 981 games.
Team 1113 played 937 games.
Team 1114 played 910 games.
Team 1115 played 520 games.
Team 1116 played 980 games.
Team 1117 played 891 games.
Team 1118 played 27 games.
Team 1119 played 868 games.
Team 1120 played 936 games.
Team 1121 played 175 games.
Team 1122 played 903 games.
Team 1123 played 906 games.
Team 1124 played 906 games.
Team 1125 played 506 games.
Team 1126 played 906 games.
Team 1127 played 442 games.
Team 1128 played 106 games.
Team 1129 played 907 games.
Team 1130 played 951 games.
Team 1131 played 945 games.
Team 1132 played 895 games.
Team 1133 played 949 games.
Team 1134 played 200 games.
Team 1135 played 847 games.
Team 1136 played 240 g

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [82]:
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

MSD_pr_season = df.groupby('Season')['ScoreDifference'].idxmax()


for season, game_index in MSD_pr_season.items():
    score_diff = df.loc[game_index, 'ScoreDifference']
    print(f"In season {season}, the game with the highest score difference is {game_index} with a score difference of {score_diff}.")

In season 1985, the game with the highest score difference is 236 with a score difference of 60.
In season 1986, the game with the highest score difference is 4731 with a score difference of 84.
In season 1987, the game with the highest score difference is 8240 with a score difference of 73.
In season 1988, the game with the highest score difference is 12046 with a score difference of 68.
In season 1989, the game with the highest score difference is 16677 with a score difference of 70.
In season 1990, the game with the highest score difference is 19502 with a score difference of 76.
In season 1991, the game with the highest score difference is 25161 with a score difference of 68.
In season 1992, the game with the highest score difference is 27997 with a score difference of 82.
In season 1993, the game with the highest score difference is 33858 with a score difference of 81.
In season 1994, the game with the highest score difference is 36404 with a score difference of 69.
In season 1995

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [83]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [84]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [85]:
type(df['Wscore'])

pandas.core.series.Series

In [86]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [87]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [88]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [89]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y